lingua_franca/lang/format_fa.py

```diff
@@ -0,0 +1,299 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2017 Mycroft AI Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from lingua_franca.lang.format_common import convert_to_mixed_fraction
+from lingua_franca.lang.common_data_fa import \
+    _farsiOnes, _farsiTens, _farsiHundreds, _farsiBig, _farsiSeperator, \
+    _farsiFrac, _farsiFracBig, _FRACTION_STRING_FA
+import math
+from lingua_franca.internal import lookup_variant
+from enum import IntEnum
+
+class NumberVariantFA(IntEnum):
+    CONVERSATIONAL = 0
+    FORMAL = 1
+
+lookup_number = lookup_variant({
+    "default": NumberVariantFA.CONVERSATIONAL,
+    "conversational": NumberVariantFA.CONVERSATIONAL,
+    "formal": NumberVariantFA.FORMAL,
+})
+
+def _applyVariant(text, variant):
+    if variant == NumberVariantFA.FORMAL:
+        for key, value in _formal_variant.items():
+            text = text.replace(value, key)
+    return text
+
+def _handleVariant(func):
+    @lookup_variant({
+        "default": NumberVariantFA.CONVERSATIONAL,
+        "conversational": NumberVariantFA.CONVERSATIONAL,
+        "formal": NumberVariantFA.FORMAL,
+    })
+    def wrapper(*args, **kwargs):
+        result = func(*args, **kwargs)
+        print(kwargs, result)
+        if 'variant' in kwargs:
+            return _applyVariant(result, kwargs['variant'])
+        else:
+            return result
+    return wrapper
+
+@_handleVariant
```

Comment:
im not sure i understand what you are trying to do, this should be

```
@lookup_variant({
           "default": NumberVariantFA.CONVERSATIONAL,
	       "conversational": NumberVariantFA.CONVERSATIONAL,
	        "formal": NumberVariantFA.FORMAL
        })
def nice_number_fa(number, speech=True, denominators=range(1, 21), variant=None):
```

Reply:
I tried not to have duplicate code. I wanted to create the format in the same decorator by replacing the numbers in result and not repeat this in all the outputs of all the functions. That's why I made another decorator.

Reply:
oh, i didnt realize this was used in more methods, let me look into it :+1:

you are right code should not be duplicated without a good reason

Reply:
@HKalbasi can you point me to a couple unittests where i cant test this?

just add new unittest section for testing variants, like [here](https://github.com/Softcatala/lingua-franca/blob/b2cf1919a819bb00b895c1f51f926bc0923fc4a5/test/test_format_ca.py#L343)

that way i can just run those individual tests and verify things work like they are supposed to, i only need 1 test case that should change depending on variant

Reply:
I added a test in [this commit](https://github.com/MycroftAI/lingua-franca/pull/171/commits/54ab5b45427a9dfd99fdb75de0227bf76219fbfe)

Reply:
@JarbasAI Do you solve this? What should I do?

Reply:
I fixed it